src/main/java/bio/terra/workspace/service/datareference/utils/DataReferenceValidationUtils.java

```diff
@@ -1,56 +1,67 @@
 import bio.terra.workspace.service.datareference.exception.InvalidDataReferenceException;
+import bio.terra.workspace.service.datareference.model.DataReferenceType;
+import bio.terra.workspace.service.datareference.model.ReferenceObject;
+import bio.terra.workspace.service.datareference.model.SnapshotReference;
 import bio.terra.workspace.service.datarepo.DataRepoService;
 import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
 import java.util.regex.Pattern;
 import org.apache.commons.lang3.StringUtils;
 import org.springframework.stereotype.Component;
 
+/** A collection of validation functions for data references. */
 @Component
 public class DataReferenceValidationUtils {
 
   private DataRepoService dataRepoService;
-  private final Pattern nameValidationPattern = Pattern.compile("^[a-zA-Z0-9][_a-zA-Z0-9]{0,62}$");
 
-  public DataReferenceValidationUtils(DataRepoService dataRepoService) {
-    this.dataRepoService = dataRepoService;
-  }
+  /**
+   * Names must be 1-63 characters long, and may consist of alphanumeric characters and underscores
+   * (but may not start with an underscore). These restrictions match TDR snapshot name restrictions
+   * as we often expect users to use snapshot names as reference names, though this isn't required.
+   */
+  public static final Pattern NAME_VALIDATION_PATTERN =
+      Pattern.compile("^[a-zA-Z0-9][_a-zA-Z0-9]{0,62}$");
 
-  public void validateReferenceName(String name) {
-    if (StringUtils.isEmpty(name) || !nameValidationPattern.matcher(name).matches()) {
+  public static void validateReferenceName(String name) {
+    if (StringUtils.isEmpty(name) || !NAME_VALIDATION_PATTERN.matcher(name).matches()) {
       throw new InvalidDataReferenceException(
           "Invalid reference name specified. Name must be 1 to 63 alphanumeric characters or underscores, and cannot start with an underscore.");
     }
   }
 
-  public DataRepoSnapshot validateReference(
-      ReferenceTypeEnum referenceType,
-      DataRepoSnapshot reference,
+  public DataReferenceValidationUtils(DataRepoService dataRepoService) {
+    this.dataRepoService = dataRepoService;
+  }
+
+  /**
+   * Validates a referenceObject, with specific validation rules varying based on the actual type of
+   * the object.
+   */
+  public void validateReferenceObject(
+      ReferenceObject reference,
+      DataReferenceType referenceType,
       AuthenticatedUserRequest userReq) {
 
     switch (referenceType) {
       case DATA_REPO_SNAPSHOT:
-        validateDataRepoReference(reference, userReq);
-        return reference;
+        validateSnapshotReference((SnapshotReference) reference, userReq);
+        return;
       default:
         throw new InvalidDataReferenceException(
             "Invalid reference type specified. Valid types include: "
-                + DATA_REPO_SNAPSHOT.toString());
+                + DataReferenceType.DATA_REPO_SNAPSHOT.toString());
     }
   }
```

Comment:
once we have more reference types, this can throw a ClassCastException if the `referenceType` does not match the actual type of the `reference` - for instance, where we try to cast to a `SnapshotReference`. Maybe wrap this all in a try/catch and provide an explanatory error message for this case? (fine to defer this for a future PR, once we have more types)

Reply:
Agreed, we should probably be checking if the provided type doesn't match the actual type. I'll defer until we have multiple types, but added a TODO here in the meantime.